torch_xla/__init__.py

```diff
@@ -91,15 +91,16 @@ def _tpu_vm_init():
 if int(os.environ.get('PT_XLA_DEBUG', '0')):
   _fd, _tmp_fname = _setup_debug_env()
 
-# Setup libtpu library for TPU VM.
-_tpu_vm_init()
-
 import atexit
 import torch
 from ._patched_functions import _apply_patches
 from .version import __version__
 import _XLAC
 
+# Set path for libtpu.so. Let static initializer fail during _XLAC import.
+# libtpu should be loaded when the ComputationClient is created.
+_tpu_vm_init()
```

Comment:
Does this always work? Will it just load from `/lib` regardless?

Reply:
The FindAndLoadTpuLibrary will try to load from `$TPU_LIBRARY_PATH `, defaulting to `libtpu.so` if it's not set.
 https://github.com/tensorflow/tensorflow/blob/15fde8ce16bcdb418a05059f7089ea58f1da68ca/tensorflow/core/tpu/tpu_initializer_helper.cc#L259

Checking the docs for [`dlopen`](https://man7.org/linux/man-pages/man3/dlopen.3.html), it looks like if `libtpu.so` is on the `LD_LIBRARY_PATH`, then the load will succeed. I didn't notice the issue because my development environment didn't have `/lib/libtpu.so`. I guess one alternative is we can set `TPU_LIBRARY_PATH` to some nonsense value before `_XLAC`.

Reply:
I am a bit concern that failing to load `libtpu.so` in the first attempt will leave some error log which will be very confusing to our user. I understand that updating tf can be pretty annoying.. Can we at least do this intentionlly fail only if PJRT env var is configured and leave a log to explain the load failure in the first attempt is expected?

Reply:
The TPU initialization actually errors out silently if the library does not exist or is invalid. All it prints is the value of TPU_LIBRARY_PATH. Here are the logs from XRT with `TF_CPP_MIN_LOG_LEVEL=0` and some vlogs turned on:

```
root@845104b51fc1:/workspaces/pjrt# python
Python 3.8.13 (default, Apr 20 2022, 18:59:55) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch_xla.core.xla_model as xm
2022-05-27 18:39:16.376314: I tensorflow/core/tpu/tpu_initializer_helper.cc:253] Libtpu path is: libtpu.so
>>> xm.xla_device()
2022-05-27 18:39:20.954475: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:277] XRT device (LOCAL) CPU:0 -> /job:localservice/replica:0/task:0/device:XLA_CPU:0
2022-05-27 18:39:20.954527: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:277] XRT device (LOCAL) TPU:0 -> /job:localservice/replica:0/task:0/device:TPU:0
[...]
2022-05-27 18:39:20.954640: I tensorflow/core/tpu/tpu_initializer_helper.cc:253] Libtpu path is: /home/ptxla/.local/lib/python3.8/site-packages/libtpu/libtpu.so
2022-05-27 18:39:21.036084: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:286] libtpu status: OK
2022-05-27 18:39:21.036136: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:288] XRT default device: TPU:0
[...]
2022-05-27 18:39:26.618779: I tensorflow/compiler/xla/service/service.cc:174] XLA service 0x55f8822671e0 initialized for platform TPU (this does not guarantee that XLA will be used). Devices:
2022-05-27 18:39:26.618840: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (0): TPU, 2a886c8
[...]

device(type='xla', index=1)
```

If the user doesn't have any info logs enabled for debugging, they'll never actually see any of this. What do you think of setting `TPU_LIBRARY_PATH` to something obviously invalid (e.g. `/dev/null`) and printing an info log in `__init__.py`? This should be safe for XRT and PjRt, and I would rather avoid having to interleave different configuration/initialization for XRT and PjRt as much as we can.

Reply:
`printing an info log in __init__.py` sgtm. Thanks!

Reply:
Done. No errors show up when info logs are disabled. When info logs are enabled, this is our output:

<details>

<summary> Sample info log output </summary>

XRT

```
# PJRT_DEVICE=TPU python
Python 3.8.13 (default, Apr 20 2022, 18:59:55) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch_xla.core.xla_model as xm
INFO:torch_xla:Letting libtpu.so load fail during _XLAC import. libtpu.so will beloaded from Python package when the ComputationClient is created.
2022-05-31 16:56:53.760138: I tensorflow/core/tpu/tpu_initializer_helper.cc:253] Libtpu path is: /dev/null
>>> xm.xla_device()
2022-05-31 16:56:57.889400: I tensorflow/core/tpu/tpu_initializer_helper.cc:253] Libtpu path is: /home/ptxla/.local/lib/python3.8/site-packages/libtpu/libtpu.so
2022-05-31 16:57:18.172480: I tensorflow/compiler/xla/service/service.cc:174] XLA service 0x558d708d3d30 initialized for platform TPU (this does not guarantee that XLA will be used). Devices:
2022-05-31 16:57:18.172539: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (0): TPU, 2a886c8
2022-05-31 16:57:18.172549: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (1): TPU, 2a886c8
2022-05-31 16:57:18.172562: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (2): TPU, 2a886c8
2022-05-31 16:57:18.172578: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (3): TPU, 2a886c8
2022-05-31 16:57:18.172586: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (4): TPU, 2a886c8
2022-05-31 16:57:18.172597: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (5): TPU, 2a886c8
2022-05-31 16:57:18.172608: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (6): TPU, 2a886c8
2022-05-31 16:57:18.172620: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (7): TPU, 2a886c8
device(type='xla', index=0)
```

PJRT

```
# python
>>> import torch_xla.core.xla_model as xm
INFO:torch_xla:Letting libtpu.so load fail during _XLAC import. libtpu.so will be loaded from `libtpu` Python package when the ComputationClient is created.
2022-05-31 17:11:59.806583: I tensorflow/core/tpu/tpu_initializer_helper.cc:253] Libtpu path is: /dev/null
>>> xm.xla_device()
2022-05-31 17:12:01.702188: I tensorflow/core/tpu/tpu_initializer_helper.cc:253] Libtpu path is: /home/ptxla/.local/lib/python3.8/site-packages/libtpu/libtpu.so
2022-05-31 17:12:01.804383: I tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:286] libtpu status: OK
2022-05-31 17:12:01.804526: I tensorflow/compiler/xla/xla_client/xrt_local_service.cc:40] Peer localservice 1 {localhost:51011}
2022-05-31 17:12:01.804823: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-05-31 17:12:01.853358: I tensorflow/compiler/xla/service/service.cc:174] XLA service 0x55e46288a740 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2022-05-31 17:12:01.853403: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (0): Host, Default Version
2022-05-31 17:12:15.578768: I tensorflow/compiler/xla/service/service.cc:174] XLA service 0x55e464617590 initialized for platform TPU (this does not guarantee that XLA will be used). Devices:
2022-05-31 17:12:15.578825: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (0): TPU, 2a886c8
2022-05-31 17:12:15.578835: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (1): TPU, 2a886c8
2022-05-31 17:12:15.578849: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (2): TPU, 2a886c8
2022-05-31 17:12:15.578857: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (3): TPU, 2a886c8
2022-05-31 17:12:15.578868: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (4): TPU, 2a886c8
2022-05-31 17:12:15.578879: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (5): TPU, 2a886c8
2022-05-31 17:12:15.578891: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (6): TPU, 2a886c8
2022-05-31 17:12:15.578902: I tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (7): TPU, 2a886c8
2022-05-31 17:12:15.596874: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:272] Initialize GrpcChannelCache for job localservice -> {0 -> localhost:51011}
2022-05-31 17:12:15.597281: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:438] Started server with target: grpc://localhost:51011
2022-05-31 17:12:15.600504: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:354] MLIR V1 optimization pass is not enabled
device(type='xla', index=1)
```

</details>